### PR TITLE
vapor 4: basics -> async

### PR DESCRIPTION
**Pull request by tanner0101**
Adds a new Async section to the Basics guide.

### CODE REVIEW

File: 4.0/docs/async.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+/// Block until the string is ready
+let string = try futureString.wait()
+print(string) /// String
+```
+
+`wait()` can only be used on a background thread or the main thread, i.e., in `configure.swift`. It can _not_ be used on an event loop thread, i.e., in route closures.
+
+!!! warning
+    Attempting to call `wait()` on an event loop thread will cause an assertion.
```
Comment (MrLotU) [inline 4.0/docs/async.md]: It'll cause an assertion failure, no?

File: 4.0/docs/async.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+```
+
+### Blocking
+
+Calling blocking code on an event loop thread can prevent your application from responding to incoming requests. 
+
+!!! danger
+    Never make blocking calls directly on an event loop.
+    
+An example of a blocking call would be something like `libc.sleep(_:)`.
```
Comment (mcdappdev) [inline 4.0/docs/async.md]: Maybe also add Bcrypt here?

File: 4.0/docs/async.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+Promises can be completed (`succeed` / `fail`) from any thread. This is why promises require an event loop to be initialized. Promises ensure that the completion action gets returned to its event loop for execution.
+
+## Event Loop
+
+When your application boots, it will usually create one event loop for each core in the CPU it is running on. Each event loop has exactly one thread. If you are familiar with event loops from Node.js, the ones in Vapor are similar. The main difference is that Vapor can run multiple event loops in one process since Swift supports multi-threading.
+
+Each time a client connects to your server, it will be assigned to one of the event loops. From that point on, all communication between the server and that client will happen on that same event loop (and by association, that event loop's thread). 
+
+The event loop is responsible for keeping track of each connected client's state. If there is a request from the client waiting to be read, the event loop trigger a read notification, causing the data to be read. Once the entire request is read, any futures waiting for that request's data will be completed.
```
Comment (grundoon) [inline 4.0/docs/async.md]: > the event loop trigger a read notification

the event loop triggers ...

File: 4.0/docs/async.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+The event loop is responsible for keeping track of each connected client's state. If there is a request from the client waiting to be read, the event loop trigger a read notification, causing the data to be read. Once the entire request is read, any futures waiting for that request's data will be completed. 
+
+In route closures, you can access the current event loop via `Request`. 
+
+```swift
+req.eventLoop.makePromise(of: ...)
+```
+
+!!! warning
+    Vapor expects that route closures will stay on `req.eventLoop`. If you hops threads, you must ensure access to `Request` and the final response future all happen on the request's event loop.
```
Comment (grundoon) [inline 4.0/docs/async.md]: > If you hops threads

If you hop ...

File: 4.0/docs/async.md (hunk 2 above)
Comment (tanner0101) [reply 4.0/docs/async.md]: I think CPU bound work is more of a gray area (for me at least) than what this section is referencing, i.e., specifically IO bound work. 

The implications of CPU bound work hogging event loop threads is definitely worth documenting though. I would need to do more research to write about that though.

**Pull request merged**